test: XORed blocks test follow up

tdb3 opened the pull request:
Builds on PR #30657.

Refactors `read_xor_key()` from `util.py` to `test_node.py` (comment https://github.com/bitcoin/bitcoin/pull/30657#discussion_r1723358327)

Adds a check that `xor.dat` is created when missing (comment https://github.com/bitcoin/bitcoin/pull/30657#discussion_r1717724161)

Help states:
```
-blocksxor
       Whether an XOR-key applies to blocksdir *.dat files. The created XOR-key
       will be zeros for an existing blocksdir or when `-blocksxor=0` is
       set, and random for a freshly initialized blocksdir. (default: 1)
```

danielabrozzoni reviewed (approved):
tACK 07365d026127eb8435e7919a7793bd25d7f3ceca

brunoerg commented on test/functional/feature_blocksxor.py:
```diff
@@ -59,6 +60,8 @@ def run_test(self):
         # nblocks=0    -> verify all blocks
         node.verifychain(checklevel=2, nblocks=0)
+        self.log.info("Check that xor.dat is recreated")
+        assert read_xor_key(node=node) == NULL_BLK_XOR_KEY
```
```suggestion
        assert_equal(read_xor_key(node=node), NULL_BLK_XOR_KEY)
```

tdb3 replied:
Thanks for reviewing.  Implemented

tdb3 commented:
Thanks @danielabrozzoni and @brunoerg.  Pushed an update to address comments, and clarified the log message.

theStack reviewed (approved):
ACK 7b7c287a33dc0c6f09bf32f31a8f05941b0cc9a2

Thanks for following up! If you want, you could also tackle the suggestion to move the `read_xor_key` helper to the `TestNode` class, see https://github.com/bitcoin/bitcoin/pull/30657#discussion_r1717985952

glozow reviewed (commented):
ACK 7b7c287a33dc0c6f09bf32f31a8f05941b0cc9a2

brunoerg reviewed (approved):
ACK 7b7c287a33dc0c6f09bf32f31a8f05941b0cc9a2

tdb3 committed:
test: add blocks_key_path

Adds a convenience function to TestNode
to provide the path to the blocks xor key.
Updates util and feature_blocksxor to use it.

tdb3 committed:
refactor: move read_xor_key() to TestNode

tdb3 committed:
refactor: lift NUM_XOR_BYTES

tdb3 committed:
test: add null block xor key

tdb3 committed:
test: check xor.dat recreated when missing

tdb3 commented on test/functional/test_framework/util.py:
```diff
@@ -391,6 +391,8 @@ def rpc_url(datadir, i, chain, rpchost):
 # from InitBlockdirXorKey::xor_key.size()
 NUM_XOR_BYTES = 8
+# The null block key (all 0s)
+NULL_BLK_XOR_KEY = bytes(NUM_XOR_BYTES)
```
lgtm, added.

tdb3 commented:
> If you want, you could also tackle the suggestion to move the `read_xor_key` helper to the `TestNode` class, see [#30657 (comment)](https://github.com/bitcoin/bitcoin/pull/30657#discussion_r1717985952)

Thanks for the suggestion!  This PR seems to be a reasonable place to perform the move, so I updated it to include this. Happy to make tweaks if you or @maflcko have any further suggestions.

theStack commented on test/functional/test_framework/test_node.py:
```diff
@@ -469,6 +469,12 @@ def blocks_path(self) -> Path:
+    def read_xor_key(self) -> bytes:
+        with open(self.blocks_key_path, "rb") as xor_f:
+            NUM_XOR_BYTES = 8 # From InitBlocksdirXorKey::xor_key.size()
+            return xor_f.read(NUM_XOR_BYTES)
```
Not sure if there are guidelines for this (and if yes, how strict we are), but I would tend to avoid using the "property" decorator for non-trivial methods that access state outside of the class (in this case, by interacting with the file system). At least I wouldn't expect as user that accessing a class field ever involves any I/O. I think dropping the property and keeping the name as-is "read_xor_key" might be just fine?

tdb3 replied:
Updated

theStack commented on test/functional/feature_blocksxor.py:
```diff
@@ -54,7 +53,7 @@ def run_test(self):
             match=ErrorMatch.PARTIAL_REGEX)
 
         self.log.info("Delete XOR key, restart node with '-blocksxor=0', check blk*.dat/rev*.dat file integrity")
-        os.remove(node.blocks_path / 'xor.dat')
+        os.remove(node.blocks_key_path)
```
nit: could put this already into the commit that introduces the blocks_key_path property to TestNode.

tdb3 replied:
Yup, this is more coherent.  Updated.

theStack reviewed (approved):
ACK ab9d2451c47fd30abecaea3f8d8a6e33541911d7

tdb3 commented on test/functional/test_framework/test_node.py:
```diff
@@ -43,6 +43,9 @@
 
 BITCOIND_PROC_WAIT_TIMEOUT = 60
+# The size of the blocks xor key
+# from InitBlockdirXorKey::xor_key.size()
```
Thanks.  Fixed.

theStack reviewed (approved):
re-ACK 1c403af92be13cbdfa6ab1906f8cc34fabc21e63

tdb3 committed:
refactor: use unlink rather than os.remove

tdb3 commented on test/functional/feature_blocksxor.py:
```diff
@@ -54,11 +56,13 @@ def run_test(self):
 
         self.log.info("Delete XOR key, restart node with '-blocksxor=0', check blk*.dat/rev*.dat file integrity")
-        os.remove(node.blocks_path / 'xor.dat')
+        os.remove(node.blocks_key_path)
```
Thanks, that's cleaner.  Added.

theStack reviewed (approved):
re-ACK e1d5dd732d5dc641faf1dde316275c84b6bb224b

brunoerg reviewed (approved):
reACK e1d5dd732d5dc641faf1dde316275c84b6bb224b

achow101 commented:
ACK e1d5dd732d5dc641faf1dde316275c84b6bb224b
